lib/pbench/test/functional/server/test_put.py

```diff
@@ -35,44 +39,61 @@ def test_upload_all(self, server_client: PbenchServerClient, login_user):
         datasets = server_client.get_list(
             metadata=["dataset.access", "server.tarball-path", "server.status"]
         )
-        found = sorted(d.name for d in datasets)
-        expected = sorted(self.tarballs.keys())
-        assert found == expected
-        for dataset in datasets:
-            t = self.tarballs[dataset.name]
-            assert dataset.name in dataset.metadata["server.tarball-path"]
-            assert dataset.metadata["server.status"]["upload"] == "ok"
-            assert t.access == dataset.metadata["dataset.access"]
+        found = {d.name for d in datasets}
+        expected = frozenset(self.tarballs.keys())
+        assert expected.issubset(found), f"expected {expected!r}, found {found!r}"
```

Comment:
If we're really going to try to ignore pre-existing datasets I'd prefer to do it rigorously; e.g., by capturing a full list *before* the PUT test and making sure that what we see *after* the PUT test is the ones we added plus the ones we found earlier.

Unfortunately, if we were to run against an operational server, there's a risk that someone will push additional datasets while we're running... on the other hand, I'm not sure I'm entirely comfortable with the idea of doing that.

Reply:
These questions are what we need to decide with regard to these tests.

 * Do we want these tests to assume a pristine system or not?
 * Do we want to allow other actors to modify the pbench server during the test run?

My vote is that we don't assume a pristine system ahead of time, and we want to allow other actors to engage in changes to the pbench server during the test run.

For functional tests, I don't see how we'd gain enough value from the tests to warrant enforcing such constraints.

If somebody can propose a test that exercises our APIs that require a pristine environment to ensure the API works correctly, then it should be proposed as a separate change.

Moving forward under a non-pristine environment, the test uses the `upload` API and the `datasets_list` API.  To verify we successfully uploaded the list of data sets, we need to:

 1. Verify the `upload` API returned success
 1. Use the `datasets_list` API to return a list of data sets our user has access to and verify that each of the uploaded data sets are in that list

That is the change we made here: just check that the uploaded data sets are in the available list of data sets.

Reply:
The primary environment is the CI pipeline. Assuming a pristine environment allows a more thorough check of the context. (For example, that we're not reporting phantom datasets somehow, which this code would just ignore.) We don't have complete tests unless we fully control the environment.

I can see a use case for a limited suite that can be run against a live server, but I think we need to be really careful about a lot more things.

Reply:
> To verify we successfully uploaded the list of data sets, we need to:
>
> 1. Verify the upload API returned success
> 1. Use the datasets_list API to return a list of data sets our user has access to and verify that each of the uploaded data sets are in that list

If I might be so bold...we should probably determine, before step 1, whether any of the datasets that we intend to upload are already there...otherwise, we cannot properly evaluate the "success" of the upload in step 1....

That is, the idea of the test is to do something to the server and then verify that the Server state has changed appropriately (as well as that the operations' results were reported back correctly); previously, we could omit the "step 0" check, because we knew (er, assumed...) that the Server was in a pristine state...now that we're waiving that assumption, we need the additional check.  So, I concur with Dave:  the test should collect the Server state prior to the test, collect the state again after the test, and then ensure that the differences between the two are not inconsistent with the operations executed by the test...which sounds only slightly challenging.

Reply:
I believe the `PUT` APIs fail if the data set has already been uploaded.

Reply:
> I believe the `PUT` APIs fail if the data set has already been uploaded.

No; in fact the new duplicate upload test I'm trying to rebase on top of the NGINX changes (when the container is actually available because I'm too lazy to try to build it from scratch) checks the distinction. The first PUT returns 201 (CREATED) while a true duplicate (matching MD5 and name) returns 200 (OK).

The "current" test only checks `response.ok` which allows any 200. The new tests will explicitly look for 201 on the normal PUT and then 200 on a new duplicate upload test.

Reply:
Well that is non-intuitive.  So we'll allow somebody to upload a N Gbs of data again?

We should be check the MD5 header field and failing if the exact same tar ball would be uploaded without having to ingest all that data.

I'll file something for that.

Reply:
I believe that we put some thought behind this and decided that it should _not_ be an error for someone to re-upload a file (based on a matching hash).

Now that said, we don't necessarily have to wait to receive all N Gbs of the data before we tell them "you're good, we've got that one"...right?  Can't the Server work with the proxy to get the headers (including the hash) ahead of the payload?

With that, the `200` vs. `201` properly captures the semantics, and the rest of the implementation has flexibility act appropriately.

Reply:
Okay, it turns out we are a bit more confusing.  Before we start reading the body of the PUT request, we return `OK`.  Yikes.  So a client will consider that success that we sent a 200 but never read the payload?

Reply:
Withdrew PR #3150.